src/jinja2/utils.py

```diff
@@ -593,17 +598,56 @@ def select_autoescape(
     If nothing matches then the initial value of autoescaping is set to the
     value of `default`.
 
+    The `special_extensions` is a dictionary which keys are the extensions
+    to be considered and the values are the escape function to be used
+    to escape this kind of files.
+
+    I.e. if you use the `latex package <https://pypi.org/project/latex/>`_,
+    you can create an environment that escapes all LaTeX files
+    with the correct escaper but still handle HTML files correctly::
+
+        from jinja2 import Environment, select_autoescape
+        from latex import escape as latex_escape
+        env = Environment(autoescape=select_autoescape(
+            special_extensions={'tex': latex_escape}
+        ))
+
     For security reasons this function operates case insensitive.
 
     .. versionadded:: 2.9
+        created function
+    .. versionchanged:: 3.0
+        parameter ``special_extensions`` was added
     """
-    enabled_patterns = tuple(f".{x.lstrip('.').lower()}" for x in enabled_extensions)
-    disabled_patterns = tuple(f".{x.lstrip('.').lower()}" for x in disabled_extensions)
+
+    def extension_str(x):
+        """return a lower case extension always starting with point"""
+        return f".{x.lstrip('.').lower()}"
+
+    enabled_patterns = tuple(extension_str(x) for x in enabled_extensions)
+    disabled_patterns = tuple(extension_str(x) for x in disabled_extensions)
+
+    if special_extensions is None:
+        special_extensions = {}
+    if special_extensions is False:
+        special_extensions = {}
+    special_extensions = {
+        extension_str(key): func for key, func in special_extensions.items()
+    }
 
     def autoescape(template_name):
         if template_name is None:
             return default_for_string
         template_name = template_name.lower()
+        # create dictionary so that have lower case keys sorted from longest
+        # to shortest, so we can easily find the best match
+        # i.e. if there are .la.tex and .tex endings in the dictionary,
+        # .la.tex ending should be found if the filename if .la.tex is given
+        for key, func in sorted(
+            special_extensions.items(), key=lambda x: len(x[0]), reverse=True
+        ):
```

Comment:
The commend above is overly verbose IMO. Something like "Lookup autoescape function using the longest matching suffix" explains it just as well.

That aside, do we really need this complexity? The only file type where double extensions are somewhat common are archives (`.tar.{gz,xz,bz2}`). So I'd rather just match using the actual extension (as determined by `os.path.splitext`) instaed of making this more complex than necessary.

Reply:
Well I can imagine end users that want a custom escape function for `.special.html` and also `i.dont.like.spaces.but.points.html`

Reply:
What would be the use case for this? I get `.html` vs `.tex`, but anything `.html` should probably use the standard HTML escaper...

Reply:
People could (arguably maybe misuse) use the escape function to escape all bad words like "fuck" from endusers.
Or people use a javascript framework that replaces a specific part of the HTML, so the end users should never be allow to write this special string.

IMO libraries should be enablers and if I see a potential bug/edge case I rather fix it right away as frustrate an end user/library user in the end.

Reply:
> arguably maybe misuse

indeed, which is why I'm -1 on this

> Or people use a javascript framework that replaces a specific part of the HTML, so the end users should never be allow to write this special string.
> IMO libraries should be enablers and if I see a potential bug/edge case I rather fix it right away as frustrate an end user/library user in the end.

I get your point, but when people go into *that* kind of obscure territory they should probably provide their own `select_autoescape` implementation... In the end, libraries/framework should do their main jobs well, and if someone wants to do obscure things they can do that by subclassing/overriding things (or in this case providing their own autoescape function)

Reply:
Sure but honestly I don't know any better solution :dizzy:. 
Using splitext simply *feels* wrong for me:
```
from os.path import splitext
splitext('foobar.ver1.html')
('foobar.ver1', '.html')
splitext('foobar.gz.html')
('foobar.gz', '.html')
splitext('foobar.tar.gz')
('foobar.tar', '.gz')
```

So please provide a change request with your preferred solution . Already spend much more time on this PR then intended and really would not like to invest time implementing something that *feels* like implementing a 🐞 bug  for me. :lotus_position: